tests/common.py

```diff
@@ -0,0 +1,43 @@
+from __future__ import annotations
+
+from abc import ABC
+
+import yaml
+
+from mewbot.loader import load_component
+from mewbot.core import ComponentKind
+from mewbot.component import ConfigBlock
+from mewbot.io.http import HTTPServlet
+
+
+class BaseTestClassWithConfig(ABC):
+    # pylint: disable=R0903
+    # Too few public methods, because this is always used elsewhere
+    fname: str
+    config: ConfigBlock
+    doctype: ComponentKind
+    config_toload = True
+
```

Comment:
I love the concept of this class! Very nice way to load a component from YAML.

This comment covers two additional thoughts:
 - I find overriding properties to be an unclear way to give info to a super class
 - We could do some fun things with the type system

For the first, I'd recommend setting up some stuff in `__init__`:
```suggestion
class BaseTestClassWithConfig(ABC):
    """Caches configs and components for testing"""
    fname: str
    doctype: ComponentKind
    _config: Optional[ConfigBlock] = None
    config_toload = True

    def __init__(self, filename: str, kind: ComponentKind) -> None:
        self.fname = filename
        self.doctype = kind

    @property
    def config(self) -> ConfigBlock:
        self.cache_config()
        return self._config

    @property
    def component(self) -> Component:
        self.cache_config()
        return load_component(self.config)
```

For `ExampleHttpPostIOConfig`, this would make the class look more like:
```python
class ExampleHttpPostIOConfig(BaseTestClassWithConfig):
    def __init__(self) -> None:
        super().__init__("examples/trivial_http_post.yaml", ComponentKind.IO_CONFIG)

    @property
    def component(self) -> HTTPServlet:
        component = super().component
        assert isinstance(component, HTTPServlet)
        return component
```

The typing idea in the second bullet is around using the Generics in python typing, which might look something like:

```suggestion
T = TypeVar("T", bound=Component, covariant=True)


class BaseTestClassWithConfig(ABC, Generic[T]):
    _config: ConfigBlock
    _component: Optional[T] = None

    def __init__(self, filename: str, implementation: Type[T]) -> None:
        with open(filename, "r", encoding="utf-8") as config:
            self._config = next([document for document in yaml.load_all(config, Loader=yaml.CSafeLoader) if document["implement"] == implementation.__name__])

    @property
    def config(self) -> ConfigBlock:
        return self._config

    @property
    def component(self) -> T:
        if not self._component:
            component = load_component(self._config)
            assert isinstance(component, T), "Config loaded incorrect type"
            self._component = component
        return self._component
```

Which should mean that doing something like `BaseTestClassWithConfig("examples/my_example.yaml", mewbot.discord.DiscordIO)` will type check as returns in a `DiscordIO` without any extra work

Reply:
^ Note: I wrote all that from memory, bug me if things don't actually work

Reply:
I second the above on the form - very nice!

Reply:
So it turns out that `pytest` will ignore classes with `__init__`, but I have been able to do "load this specific object from a config file" which looks like

```python
class TestLoaderHttpsPost(BaseTestClassWithConfig[HTTPServlet]):
    config_file: str = "examples/trivial_http_post.yaml"
    implementation: Type[HTTPServlet] = HTTPServlet

    # Test this working
    def test_working(self) -> None:
        assert self.config["implementation"] == "mewbot.io.http.HTTPServlet"

    # Test that the loading is accurate
    def test_loading_component_type(self) -> None:
        assert isinstance(self.component, HTTPServlet)
```